File: purecoincidence/src/main/scala/purecoincidence/PureExistence.scala
package purecoincidence

object PureExistence extends App {
  def divide(x: Int, y: Int): Option[Int] =
    if (y == 0) None
    else Some(x / y)

  def parseInt(s: String): Option[Int] = {
    try {
      Some(s.toInt)
    } catch {
      case _: NumberFormatException => None
    }   // == scala.util.Try(s.toInt).toOption
  }

  def stringDivide0(sx: String, sy: String): Option[Int] =
    parseInt(sx).flatMap { x =>
      parseInt(sy).flatMap { y =>
        divide(x, y)
      }
    }

  def stringDivide(sx: String, sy: String): Option[Int] =
    for {
      x <- parseInt(sx)
      y <- parseInt(sy)
      d <- divide(x, y)
    } yield d
}

File: hopefullypure/app/src/main/java/hopefullypure/PureExistence.java
package hopefullypure;

import java.util.Optional;

public class PureExistence {
    static Optional<Integer> divide(int x, int y) {
        if (y == 0) return Optional.empty();
        else return Optional.of(x / y);
    }

    static Optional<Integer> parseInt(String s) {
        try {
            return Optional.of(Integer.parseInt(s));
        } catch (NumberFormatException e) {
            return Optional.empty();
        }
    }

    static Optional<Integer> stringDivide(String sx, String sy) {
        return parseInt(sx).flatMap(x ->
                 parseInt(sy).flatMap(y ->
                   divide(x, y)));
    }
}
